compiler/qsc_eval/src/backend/noise_tests.rs

```diff
@@ -0,0 +1,181 @@
+use num_complex::Complex;
+use std::fmt::Write;
+
+#[test]
+fn pauli_noise() {
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 0.0);
+    assert!(
+        noise
+            .expect("Expected construction of noiseless noise.")
+            .is_noiseless(),
+        "Expected noiseless noise."
+    );
+    let noise = PauliNoise::from_probabilities(1e-5, 0.0, 0.0);
+    assert!(
+        !noise
+            .expect("Expected construction of 1e-5 bit flip noise.")
+            .is_noiseless(),
+        "Expected noise to be noisy."
+    );
+    let noise = PauliNoise::from_probabilities(1.0, 0.0, 0.0);
+    assert!(
+        !noise
+            .expect("Expected construction of 1.0 bit flip noise.")
+            .is_noiseless(),
+        "Expected noise to be noisy."
+    );
+    let noise = PauliNoise::from_probabilities(0.01, 0.01, 0.01)
+        .expect("Expected construction of 0.01 depolarizing noise.");
+    assert!(!noise.is_noiseless(), "Expected noise to be noisy.");
+    assert!(
+        0.0 <= noise.distribution[0]
+            && noise.distribution[0] <= noise.distribution[1]
+            && noise.distribution[1] <= noise.distribution[2]
+            && noise.distribution[2] <= 1.1,
+        "Expected non-decreasing noise distribution."
+    );
+    let _ = PauliNoise::from_probabilities(-1e-10, 0.1, 0.1)
+        .expect_err("Expected error for probabilities -1e-10, 0.1, 0.1.");
+    let _ = PauliNoise::from_probabilities(1.0 + -1e-10, 0.1, 0.1)
+        .expect_err("Expected error for probabilities 1.0+1e-10, 0.1, 0.1.");
+    let _ = PauliNoise::from_probabilities(0.3, 0.4, 0.5)
+        .expect_err("Expected error for probabilities 0.3, 0.4, 0.5.");
+}
+
+#[test]
+fn noisy_simulator() {
+    let sim = SparseSim::new();
+    assert!(sim.is_noiseless(), "Expected noiseless simulator.");
+
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 0.0).expect("Cannot construct noise.");
+    let sim = SparseSim::new_with_noise(&noise);
+    assert!(sim.is_noiseless(), "Expected noiseless simulator.");
+
+    let noise = PauliNoise::from_probabilities(1e-10, 0.0, 0.0).expect("Cannot construct noise.");
+    let sim = SparseSim::new_with_noise(&noise);
+    assert!(!sim.is_noiseless(), "Expected noisy simulator.");
+
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 1e-10).expect("Cannot construct noise.");
+    let sim = SparseSim::new_with_noise(&noise);
+    assert!(!sim.is_noiseless(), "Expected noisy simulator.");
+}
+
+#[test]
+fn noiseless_gate() {
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 0.0).expect("Cannot construct noise.");
+    let mut sim = SparseSim::new_with_noise(&noise);
+    let q = sim.qubit_allocate();
+    for _ in 0..100 {
+        sim.x(q);
+        let res1 = sim.m(q);
+        assert!(res1, "Unexpected value. True expected without noise.");
+        sim.x(q);
+        let res2 = sim.m(q);
+        assert!(!res2, "Unexpected value. False expected without noise.");
+    }
```

Comment:
I don't think a loop is needed here, I would just verify that basic bit flipping works to one and then back to zero rather than put this in a loop.

Reply:
I changed code and moved 100% bitflip "gate test" to be like other "noisy_via_?" tests without the loop. However, I kept loops in the two other tests. They need loops to test probabilistic nature of noise. First of all, if the implementation in the SparseSim is correct, all iterations should be successful. So, this won't be a flaky or "may fail sometimes" test. It is also very fast - negligible time on my computer. However, if the implementation in the SparseSim is not correct, it may produce an error, and it may do so only sometimes. The more iterations I run, the more chances that the error will reveal itself. So doing this in a loop would make test less valuable.